design-proposals/app-orch-deploy-applications.md

```diff
@@ -0,0 +1,208 @@
+
+  - `default_namesspaces`. Used to pin applications to namespaces.
+
+  - `namespaces`. Used to create and label namespaces at deployment time.
+
+In Orchestrator 3.0, a single Helm Chart application requires all three objects to be created,
+usually in a 1:1 correspondence. The user has to create the `Registry`, then they create an
+`Application`, linking it to the `Registry`. Finally, the user creates the `Deployment Package`
+and links it the `Application`. There is complexity and overhead in creating the
+`deployment_profiles` as they have to shadow the application profiles.
+There is complexity in versioning all of these objects -- updating the
+Helm Chart trickles up the chain requiring version increments to the `Application` and
+`Deployment Package`.
+
+### Deploying Applications
+
+We will extend the Orchestrator to allow deploying an `Application`, without creating a
+`Deployment Package`. To do this, we need to first make sure the `Application` contains all the
+necessary information for deployment.
+
+There are only three pieces of information in the `Deployment Package` that are not part of
+the `Application` object.
+
+- `is_deployed`. This must be added to `Application`, so we can track the deployment
+  status of an `Application`.
+
+- `default_namespaces`, `namespaces`. These can be represented as a single `default_namespace`
+  field in the `Application`. The two fields in the `Deployment Package` can remain as
+  overrides to alter the default behavior of the `Application` for more complex multi-application
+  scenarios.
+
+### Folding Registries into Applications
+
+The common case is that a Helm Chart can be retrieved from its OCI URL, and optional username
+and password. We could eliminate the need to have a separate `Registry` object by
+adding these three fields to the `Application` object:
+
+- `chart_url`. The URL of the OCI helm chart.
+
+- `chart_username`, `chart_password`. Optional credentials to use when fetching the Helm Chart.
+
+`chart_url` and the existing field `helm_registry_name` are mutually exclusive. The user must either
+choose the simplified approach, or choose the more flexible approach of using the `Registry` object.
+If `chart_url` is specified, then App Orch will also make it available as an image registry,
+consistent with how `Image Pull Secrets` are implemented in the deployment manager. As such,
+`chart_url` will also be mutually exclusive with `image_registry_name`.
+
+If the user has a more complex use case that requires a custom cert, or they want to abstract
+these details behind a named `Registry` object, they are free to do so. There is still value in
+the `Registry` object, particularly in the case where private registries with authentication
+are used as it allows sensitive security information to be stored in one object rather than
+being replicated in many. Folding registry into Application is an optional means of simplifying
+the common case, particularly for public charts.
```

Comment:
If we are `simplifying the common case, particularly for public charts`, do we need `chart_username` and `chart_password` at all?

Reply:
One of my concerns is docker throttling, which impacts a lot of the "public" applications we encounter. Being able to optionally insert a Docker PAT here might be useful.